### PR TITLE
Added IDEntityCadastralNumber

### PR DESCRIPTION
**Pull request by eldario**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes

## Description

Entity `IDEntityCadastralNumber` is added to the cadastral number of the property

Fixes # (issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code
- [x] I have made changes in [CHANGELOG.md](https://github.com/avto-dev/identity-laravel/blob/master/CHANGELOG.md) file

### CODE REVIEW

**Review by tarampampam — CHANGES_REQUESTED** [review]
Latest version is 4. You must update your branch

**Comment by eldario** [comment]
- [x] `git fetch upstream`
- [x] `git rebase upstream/master`

File: src/IDEntityInterface.php [hunk 1]
```diff
@@ -69,6 +69,13 @@ interface IDEntityInterface
      */
     public const ID_TYPE_DRIVER_LICENSE_NUMBER = 'DLN';
 
+    /**
+     * Тип - кадастровый номер (уникальный номер объекта недвижимости).
+     *
+     * @var string
+     */
+    const ID_TYPE_CADASTRAL_NUMBER = 'CAD';
```
Comment (tarampampam) [inline src/IDEntityInterface.php]: not `CADASTRAL_NUMBER`?

File: README.md [hunk 2]
```diff
@@ -53,6 +53,7 @@ $ ./artisan vendor:publish --provider="AvtoDev\IDEntity\ServiceProvider"
 `BODY`    | Номер кузова транспортного средства                            | `\AvtoDev\IDEntity\Types\IDEntityBody`
 `CHASSIS` | Номер шасси транспортного средства                             | `\AvtoDev\IDEntity\Types\IDEntityChassis`
 `DLN`     | Номер водительского удостоверения                              | `\AvtoDev\IDEntity\Types\IDEntityDriverLicenseNumber`
+`CAD    ` | Кадастровый номер объекта недвижимости                         | `\AvtoDev\IDEntity\Types\IDEntityCadastralNumber`
```
Comment (tarampampam) [inline README.md]: `$ composer require avto-dev/identity-laravel "^4.1"`

File: src/Types/IDEntityCadastralNumber.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+declare(strict_types = 1);
+
+namespace AvtoDev\IDEntity\Types;
+
+use Exception;
+use AvtoDev\ExtendedLaravelValidator\Extensions\CadastralNumberValidatorExtension;
+
+/**
+ * Идентификатор - кадастровый номер
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]: del?

File: src/Types/IDEntityCadastralNumber.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public static function normalize($value): ?string
+    {
+        try {
+            // Удаляем все символы, кроме разрешенных
+            $value = \preg_replace('~[^\d\:]~u', '', (string) $value);
+
+            //Удаляем первый символ, если это не цифра
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]: Add ` ` after `//`

File: src/Types/IDEntityCadastralNumber.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function normalize($value): ?string
+    {
+        try {
+            // Удаляем все символы, кроме разрешенных
+            $value = \preg_replace('~[^\d\:]~u', '', (string) $value);
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]: `\preg_replace` can return `string|array|null`

Comment (tarampampam) [reply src/Types/IDEntityCadastralNumber.php]: Add casting

File: src/Types/IDEntityCadastralNumber.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+            //Удаляем первый символ, если это не цифра
+            return \preg_replace('~^([^\d]+)~u', '', (string) $value);
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]: add `trim($value, ':')` after?

File: tests/Types/IDEntityCadastralNumberTest.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+declare(strict_types = 1);
+
+namespace AvtoDev\IDEntity\Tests\Types;
+
+use AvtoDev\IDEntity\IDEntity;
+use AvtoDev\IDEntity\Types\IDEntityCadastralNumber;
+
+/**
+ * Class IDEntityCadastralNumberTest.
```
Comment (tarampampam) [inline tests/Types/IDEntityCadastralNumberTest.php]: del

File: tests/Types/IDEntityCadastralNumberTest.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            '77:01:100500:1000',
+            '02:00:456789:736',
+            '66:41:1234567:81545',
+            '66:41:123456:102360',
+        ];
+
+        foreach ($valid as $value) {
+            $this->assertTrue($this->instance->setValue($value)->isValid());
+        }
+
```
Comment (tarampampam) [inline tests/Types/IDEntityCadastralNumberTest.php:69]: add more invalid values

File: tests/Types/IDEntityCadastralNumberTest.php [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        // Пробелы с двум сторон
+        $this->assertEquals($valid, $instance::normalize(' ' . $valid . ' '));
+
+        // Запрещенные символы
+        $this->assertEquals($valid, $instance::normalize('6+6:/4$1:;0(1%^0)&5*-0!0@1#:=?3'));
+
+        // С буквами
+        $this->assertEquals($valid, $instance::normalize('Start6Шесть6:4One1:01ZeRO05001:ThrEE3'));
+        //Первый символ не цифра
+        $this->assertTrue($instance->setValue(':x:y:ц:61:41:123456:102360')->isValid());
```
Comment (tarampampam) [inline tests/Types/IDEntityCadastralNumberTest.php]: This is correct? Too large value changes, I think

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: src/Types/IDEntityCadastralNumber.php (hunk 4 above)
Comment (tarampampam) [reply src/Types/IDEntityCadastralNumber.php]: + add in IDEntity provider supports for CadastralNumber

File: tests/Types/IDEntityCadastralNumberTest.php (hunk 8 above)
Comment (eldario) [reply tests/Types/IDEntityCadastralNumberTest.php]: Maybe *valid* values?

File: README.md (hunk 2 above)
Comment (eldario) [reply README.md]: Maybe I should do this *sh\*t* in `faker-providers`

File: src/IDEntityInterface.php (hunk 1 above)
Comment (eldario) [reply src/IDEntityInterface.php]: `fixed` to `CADNUM`

File: src/Types/IDEntityCadastralNumber.php (hunk 3 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: `removed`

File: src/Types/IDEntityCadastralNumber.php (hunk 4 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: - `added` single space in comment line
- added in IDEntity

File: src/Types/IDEntityCadastralNumber.php (hunk 5 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: - `added` casting in `\preg_replace`
- `changed` second `\preg_replace`  to `\trim`

File: src/Types/IDEntityCadastralNumber.php (hunk 6 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: `added`

File: tests/Types/IDEntityCadastralNumberTest.php (hunk 7 above)
Comment (eldario) [reply tests/Types/IDEntityCadastralNumberTest.php]: `removed`

File: tests/Types/IDEntityCadastralNumberTest.php (hunk 8 above)
Comment (eldario) [reply tests/Types/IDEntityCadastralNumberTest.php:69]: In the same test method:

1. `added` more **valid** values
2. `added` more **invalid** values

File: tests/Types/IDEntityCadastralNumberTest.php (hunk 9 above)
Comment (eldario) [reply tests/Types/IDEntityCadastralNumberTest.php]: `reduced string length`

**Review by tarampampam — CHANGES_REQUESTED** [review]
Told about changes

**Comment by eldario** [comment]
1. `Added` check of regional codes for each number
2. `Changed` test with **valid** & **invalid** numbers
3. `Removed` function `\trim` in `static normalize`

File: src/Types/IDEntityCadastralNumber.php [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    public static function normalize($value): ?string
+    {
+        try {
+            // Удаляем все символы, кроме разрешенных (цифры и знак ":")
+            return (string) \preg_replace('~[^\d\:]~u', '', (string) $value);
+        } catch (Exception $e) {
+            return null;
+        }
+    }
+
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]:
```php
pubf getCadNumberInfo: CadNumberInfo
{
    return new CadNumberInfo($this->value);
}
```

File: src/Types/IDEntityCadastralNumber.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * Return parsed fragments of cadastral number.
+     *
+     * @return CadastralNumberInfo
+     */
+    public function getCadastralNumberInfo(): CadastralNumberInfo
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]: `getCadastralNumberInfo` -> `getNumberInfo`

File: src/Types/IDEntityCadastralNumber.php [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        static $regions = null;
+
+        if (! $regions instanceof CadastralRegions) {
+            $regions = new CadastralRegions;
+        }
+
+        $cadastral_number_info = $this->getCadastralNumberInfo();
+
+        return $regions->getRegionByCode($cadastral_number_info->getRegionCode());
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]: `return $regions->getRegionByCode($this->getCadastralNumberInfo()->getRegionCode());`

File: src/Types/IDEntityCadastralNumber.php [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        $validator = static::getContainer()->make(CadastralNumberValidatorExtension::class);
+
+        $validated = \is_string($this->value) && $validator->passes('', $this->value);
+
+        $region_data           = $this->getRegionData();
+        $cadastral_number_info = $this->getCadastralNumberInfo();
+
+        return $validated
+               && $region_data instanceof CadastralRegionEntry
+               && $region_data->getDistricts()->hasDistrictCode($cadastral_number_info->getDistrictCode());
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php]:
```php
$region_data->getDistricts()->hasDistrictCode($this->getCadastralNumberInfo()->getDistrictCode());
```

File: src/Types/IDEntityCadastralNumber.php (hunk 11 above)
Comment (tarampampam) [reply src/Types/IDEntityCadastralNumber.php]: + add tests

File: src/Types/IDEntityCadastralNumber.php [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    public function getCadastralNumberInfo(): CadastralNumberInfo
+    {
+        return CadastralNumberInfo::parse($this->value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRegionData(): ?CadastralRegionEntry
```
Comment (tarampampam) [inline src/Types/IDEntityCadastralNumber.php:49]: + add tests

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: src/Types/IDEntityCadastralNumber.php (hunk 14 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php:49]: `modified` this test
`changed` `testNormalize` test

File: src/Types/IDEntityCadastralNumber.php (hunk 13 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: `changed`

File: src/Types/IDEntityCadastralNumber.php (hunk 12 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: `changed`

File: src/Types/IDEntityCadastralNumber.php (hunk 11 above)
Comment (eldario) [reply src/Types/IDEntityCadastralNumber.php]: 1. `changed`
2. `added`

**Review by tarampampam — APPROVED** [review]

**Comment by eldario** [comment]
> my goodness

**Pull request merged**